### PR TITLE
SQL support for t-digest based sketch aggregators

### PR DESCRIPTION
**Pull request by samarthjain**
### Description

This PR adds support for t-digest based sketch aggregators that was added in https://github.com/apache/incubator-druid/pull/7331

Additionally this PR removes previously added mergeTDigestSketch aggregator. The merging/combining functionality has been added in buildTDigestSketch aggregator. The docs also have been updated with the relevant changes.

Note that a couple of tests added in this PR will fail till https://github.com/apache/incubator-druid/pull/8099 is merged.

This PR has:
- [X ] been self-reviewed.
   - [X ] using the [concurrency checklist](https://github.com/apache/incubator-druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ X] added documentation for new or modified features or behaviors.
- [ X] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [X ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ X] added unit tests or modified existing tests to cover new code paths.
- [ ] added integration tests.
- [ X] been tested in a test Druid cluster.

For reviewers: the key changed/added classes in this PR are `TDigestGenerateSketchSqlAggregator`,  and `TDigestSketchQuantileSqlAggregator`.

### CODE REVIEW

**Comment by gianm** [comment]
Hey @samarthjain, I see this error in CI,

```
Caused by: java.lang.ClassNotFoundException: org.apache.druid.query.expression.LookupEnabledTestExprMacroTable
	at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:349)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 28 more
```

You probably need this in your pom:

```
        <dependency>
            <groupId>org.apache.druid</groupId>
            <artifactId>druid-server</artifactId>
            <version>${project.parent.version}</version>
            <scope>test</scope>
        </dependency>
```

**Comment by samarthjain** [comment]
Hmm, I do have that dependency added.
https://github.com/apache/incubator-druid/blob/adeac28b1e774990ba5fad8a52e2d7d3abe7160c/extensions-contrib/tdigestsketch/pom.xml#L157

Taking a closer look.

**Comment by gianm** [comment]
Oh, actually I think you need this:

```
        <dependency>
            <groupId>org.apache.druid</groupId>
            <artifactId>druid-server</artifactId>
            <version>${project.parent.version}</version>
            <type>test-jar</type>
            <scope>test</scope>
        </dependency>
```

(The `<type>test-jar</type>` part in particular)

**Comment by jon-wei** [comment]
@samarthjain Can you fix conflicts?

File: docs/content/development/extensions-contrib/tdigestsketch-quantiles.md [hunk 1]
```diff
@@ -70,49 +68,22 @@ Example:
...
 ```json
 {
-  "type" : "mergeTDigestSketch",
-  "name" : <output_name>,
-  "fieldName" : <metric_name>,
-  "compression": <parameter that controls size and accuracy>
- }
+	"type": "buildTDigestSketch",
+	"name": "combined_sketch",
+	"fieldName": "ingested_sketch",
```
Comment (jon-wei) [inline docs/content/development/extensions-contrib/tdigestsketch-quantiles.md]: Suggest changing `"ingested_sketch"` here to `<input-column>`, since the aggregator can accept raw values or pre-generated sketches

File: docs/content/development/extensions-contrib/tdigestsketch-quantiles.md [hunk 2]
```diff
@@ -29,16 +29,13 @@ T-Digest (https://github.com/tdunning/t-digest) is a popular datastructure for a
...
-generate sketches, just that one would be building these sketches on every query execution instead of building them
-once during ingestion.
-2. mergeTDigestSketch - used for merging pre-built T-Digest sketches. This aggregator is generally used during
-query time to combine sketches generated by buildTDigestSketch aggregator.
-3. quantilesFromTDigestSketch - used for generating quantiles from T-Digest sketches. This aggregator is generally used
-during query time to generate quantiles from sketches built using the above two sketch generating aggregators.
+1. buildTDigestSketch - It generally makes sense to use this aggregator when ingesting raw data into Druid to
+generate pre-agrgegated sketches. One can also use this aggregator during query time too to combine ingested T-Digest
+sketches.
+2. quantilesFromTDigestSketch - used for generating quantiles from T-Digest sketches. This aggregator is generally used
```
Comment (jon-wei) [inline docs/content/development/extensions-contrib/tdigestsketch-quantiles.md]: Suggest rewording this section, there's only one aggregator since the PostAggregators don't really aggregate (they only operate on values within a single row), they're more like "post aggregation transforms".

File: docs/content/development/extensions-contrib/tdigestsketch-quantiles.md [hunk 3]
```diff
@@ -133,7 +104,7 @@ This returns an array of quantiles corresponding to a given array of fractions.
 |--------|-----------|---------|
 |type|This String should always be "quantilesFromTDigestSketch"|yes|
 |name|A String for the output (result) name of the calculation.|yes|
-|fieldName|A String for the name of the input field containing raw numeric values.|yes|
+|field|A field reference pointing to the field aggregated/combined T-Digest sketch.|yes|
```
Comment (jon-wei) [inline docs/content/development/extensions-contrib/tdigestsketch-quantiles.md]: Suggest rewriting to:

"The input field reference, which must be a PostAggregator that outputs T-Digest sketches. This can be a `fieldAccess` PostAggregator, which simply returns the value of the referenced input column."

File: docs/content/development/extensions-contrib/tdigestsketch-quantiles.md [hunk 4]
```diff
@@ -162,3 +133,21 @@ Example:
...
+  "type"  : "quantileFromTDigestSketch",
+  "name": <output name>,
+  "field"  : <post aggregator that refers to a TDigestSketch (fieldAccess or another post aggregator)>,
+  "fraction" : <value>
+}
+```
+
+|property|description|required?|
+|--------|-----------|---------|
+|type|This String should always be "quantilesFromTDigestSketch"|yes|
```
Comment (jon-wei) [inline docs/content/development/extensions-contrib/tdigestsketch-quantiles.md]: quantilesFromTDigestSketch -> quantileFromTDigestSketch

File: docs/content/development/extensions-contrib/tdigestsketch-quantiles.md [hunk 5]
```diff
@@ -70,49 +68,22 @@ Example:
...
-The result of the aggregation is a T-Digest sketch that is built by merging pre-built T-Digest sketches.
-
 ```json
 {
-  "type" : "mergeTDigestSketch",
-  "name" : <output_name>,
-  "fieldName" : <metric_name>,
-  "compression": <parameter that controls size and accuracy>
- }
+	"type": "buildTDigestSketch",
```
Comment (jon-wei) [inline docs/content/development/extensions-contrib/tdigestsketch-quantiles.md]: Since build/merge implementations are combined, suggest calling the type "tDigestSketch"

File: extensions-contrib/tdigestsketch/src/main/java/org/apache/druid/query/aggregation/tdigestsketch/sql/TDigestGenerateSketchSqlAggregator.java [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+        rowSignature,
+        inputOperand
+    );
+    if (input == null) {
+      return null;
+    }
+
+    final AggregatorFactory aggregatorFactory;
+    final String aggName = StringUtils.format("%s:agg", name);
+    final RexNode maxNumEntriesOperand = Expressions.fromFieldAccess(
```
Comment (jon-wei) [inline extensions-contrib/tdigestsketch/src/main/java/org/apache/druid/query/aggregation/tdigestsketch/sql/TDigestGenerateSketchSqlAggregator.java]: `maxNumEntriesOperand` should be renamed to `compressionOperand`

File: extensions-contrib/tdigestsketch/src/main/java/org/apache/druid/query/aggregation/tdigestsketch/sql/TDigestGenerateSketchSqlAggregator.java [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+        rowSignature,
+        project,
+        aggregateCall.getArgList().get(1)
```
Comment (jon-wei) [inline extensions-contrib/tdigestsketch/src/main/java/org/apache/druid/query/aggregation/tdigestsketch/sql/TDigestGenerateSketchSqlAggregator.java]: Since the compression operand is optional, there should be a check that ` aggregateCall.getArgList()` has size > 1

File: extensions-contrib/tdigestsketch/src/main/java/org/apache/druid/query/aggregation/tdigestsketch/sql/TDigestGenerateSketchSqlAggregator.java [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+
+    TDigestGenerateSketchSqlAggFunction()
+    {
+      super(
+          NAME,
+          null,
+          SqlKind.OTHER_FUNCTION,
+          ReturnTypes.explicit(SqlTypeName.OTHER),
+          null,
+          OperandTypes.and(
```
Comment (jon-wei) [inline extensions-contrib/tdigestsketch/src/main/java/org/apache/druid/query/aggregation/tdigestsketch/sql/TDigestGenerateSketchSqlAggregator.java]: The agg function here should be adjusted to support the optional compression param like in the quantile version

File: extensions-contrib/tdigestsketch/src/main/java/org/apache/druid/query/aggregation/tdigestsketch/sql/TDigestSketchQuantileSqlAggregator.java [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+
+          if (virtualInput == null) {
+            inputMatches = input.isDirectColumnAccess()
+                           && input.getDirectColumn().equals(theFactory.getFieldName());
+          } else {
+            inputMatches = ((ExpressionVirtualColumn) virtualInput).getExpression()
+                                                                   .equals(input.getExpression());
+          }
+
+          if (inputMatches) {
```
Comment (jon-wei) [inline extensions-contrib/tdigestsketch/src/main/java/org/apache/druid/query/aggregation/tdigestsketch/sql/TDigestSketchQuantileSqlAggregator.java]: This should check that the compression and quantile parameters match as well

File: extensions-contrib/tdigestsketch/src/main/java/org/apache/druid/query/aggregation/tdigestsketch/sql/TDigestSketchQuantileSqlAggregator.java [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+            ),
+            quantile
+        )
+    );
+  }
+
+  private static class TDigestSketchQuantileSqlAggFunction extends SqlAggFunction
+  {
+    private static final String SIGNATURE1 = "'" + NAME + "(column, quantile)'\n";
+    private static final String SIGNATURE2 = "'" + NAME + "(column, quantile, compression)'\n";
```
Comment (jon-wei) [inline extensions-contrib/tdigestsketch/src/main/java/org/apache/druid/query/aggregation/tdigestsketch/sql/TDigestSketchQuantileSqlAggregator.java]: Can you update the docs to mention that optional compression param is supported for the quantile SQL function as well

File: extensions-contrib/tdigestsketch/src/test/java/org/apache/druid/query/aggregation/tdigestsketch/sql/TDigestSketchSqlAggregatorTest.java [hunk 11]
```diff
@@ -0,0 +1,383 @@
...
+        Iterables.getOnlyElement(queryLogHook.getRecordedQueries())
+    );
+  }
+
+  @Test
+  public void testComputingQuantileOnPreAggregatedSketch() throws Exception
+  {
+    SqlLifecycle sqlLifecycle = sqlLifecycleFactory.factorize();
+    final String sql = "SELECT\n"
+                       + "TDIGEST_QUANTILE(qsketch_m1, 0.1),\n"
```
Comment (jon-wei) [inline extensions-contrib/tdigestsketch/src/test/java/org/apache/druid/query/aggregation/tdigestsketch/sql/TDigestSketchSqlAggregatorTest.java]: Suggest adding tests with a quantile SQL agg that has the compression parameter is specified, and where the sketch build SQL agg uses default compression

File: processing/src/main/java/org/apache/druid/query/aggregation/AggregatorUtil.java [hunk 12]
```diff
@@ -121,8 +121,8 @@
 
   // TDigest sketch aggregators
   public static final byte TDIGEST_BUILD_SKETCH_CACHE_TYPE_ID = 0x38;
-  public static final byte TDIGEST_MERGE_SKETCH_CACHE_TYPE_ID = 0x39;
   public static final byte TDIGEST_SKETCH_TO_QUANTILES_CACHE_TYPE_ID = 0x40;
```
Comment (jon-wei) [inline processing/src/main/java/org/apache/druid/query/aggregation/AggregatorUtil.java]: Since SKETCH_TO_QUANTILES and SKETCH_TO_QUANTILE are postaggs, their IDs should go in `PostAggregatorIds` instead

File: extensions-contrib/tdigestsketch/src/main/java/org/apache/druid/query/aggregation/tdigestsketch/sql/TDigestSketchQuantileSqlAggregator.java (hunk 9 above)
Comment (samarthjain) [reply extensions-contrib/tdigestsketch/src/main/java/org/apache/druid/query/aggregation/tdigestsketch/sql/TDigestSketchQuantileSqlAggregator.java]: Hmm, not sure how to do that. Compression is part of the TDigestSketchAggregatorFactory and quantile isn't. The quantile is used in postAggregator.

Comment (jon-wei) [reply extensions-contrib/tdigestsketch/src/main/java/org/apache/druid/query/aggregation/tdigestsketch/sql/TDigestSketchQuantileSqlAggregator.java]: Ah, it should just check compression

File: extensions-contrib/tdigestsketch/src/main/java/org/apache/druid/query/aggregation/tdigestsketch/sql/TDigestSketchQuantileSqlAggregator.java (hunk 10 above)
Comment (samarthjain) [reply extensions-contrib/tdigestsketch/src/main/java/org/apache/druid/query/aggregation/tdigestsketch/sql/TDigestSketchQuantileSqlAggregator.java]: So I thought about this a bit more and I am not sure what is the right thing to do. Consider a query like this:

`SELECT TDIGEST_QUANTILE(column1, 0.1, 100), TDIGEST_QUANTILE(column1, 0.2, 200), TDIGEST_QUANTILE(column1, 0.2, 300) FROM FOO`

This query does the aggregation to generate a sketch on column1 and then applies the post aggregator to compute quantile out of it (3 times). The problem is that the sketch is generated only once using the compression param 100 (from the first `TDIGEST_QUANTILE(column1, 0.1, 100)` and the compression params from the following calls are ignored. 

Is there a way to validate in Druid SQL that all the calls on TDIGEST_QUANTILE() for a column are using the same compression param? The other alternative is to not support compression param for this method and just use the default value of compression.

Comment (jon-wei) [reply extensions-contrib/tdigestsketch/src/main/java/org/apache/druid/query/aggregation/tdigestsketch/sql/TDigestSketchQuantileSqlAggregator.java]: > The problem is that the sketch is generated only once using the compression param 100 (from the first TDIGEST_QUANTILE(column1, 0.1, 100) and the compression params from the following calls are ignored.

If you add a check that the compression parameter matches before reusing the 
aggregator (https://github.com/apache/incubator-druid/pull/8100#discussion_r309914605), does this still occur?

Comment (samarthjain) [reply extensions-contrib/tdigestsketch/src/main/java/org/apache/druid/query/aggregation/tdigestsketch/sql/TDigestSketchQuantileSqlAggregator.java]: Thanks, that worked.
